dex/pair/src/safe_price.rs

```diff
@@ -0,0 +1,239 @@
+    fn get_tokens_for_given_position_with_safe_price(
+        &self,
+        liquidity: BigUint,
+    ) -> MultiResult2<EsdtTokenPayment<Self::Api>, EsdtTokenPayment<Self::Api>> {
+        self.update_safe_state_on_the_fly();
+
+        let c_state = self.current_state().get();
+        let total_supply = self.lp_token_supply().get();
+        let first_token_id = self.first_token_id().get();
+        let second_token_id = self.second_token_id().get();
+        let big_zero = BigUint::zero();
+
+        let (first_token_worth, second_token_worth) = if total_supply == big_zero
+            || c_state.first_token_reserve == big_zero
+            || c_state.second_token_reserve == big_zero
+        {
+            (big_zero.clone(), big_zero)
+        } else {
+            let first_worth = &liquidity * &c_state.first_token_reserve / &total_supply;
+            let second_worth = &liquidity * &c_state.second_token_reserve / &total_supply;
+
+            (first_worth, second_worth)
+        };
+
+        MultiResult2::from((
+            EsdtTokenPayment::new(first_token_id, 0, first_token_worth),
+            EsdtTokenPayment::new(second_token_id, 0, second_token_worth),
+        ))
+    }
+
+    #[endpoint(getSafePrice)]
+    fn get_safe_price(&self, input: EsdtTokenPayment<Self::Api>) -> EsdtTokenPayment<Self::Api> {
+        self.update_safe_state_on_the_fly();
+
+        let first_token_id = self.first_token_id().get();
+        let second_token_id = self.second_token_id().get();
+        let c_state = self.current_state().get();
+
+        let (r_in, r_out, t_out) = if input.token_identifier == first_token_id {
+            let r_in = c_state.first_token_reserve.clone();
+            let r_out = c_state.second_token_reserve;
+            let t_out = second_token_id;
+
+            (r_in, r_out, t_out)
+        } else if input.token_identifier == second_token_id {
+            let r_in = c_state.second_token_reserve.clone();
+            let r_out = c_state.first_token_reserve;
+            let t_out = first_token_id;
+
+            (r_in, r_out, t_out)
+        } else {
+            assert!(self, ERROR_UNKNOWN_TOKEN);
+        };
+        assert!(
+            self,
+            input.amount != 0u64 && r_in != 0u64 && r_out != 0u64,
+            ERROR_ZERO_AMOUNT,
+        );
+
+        EsdtTokenPayment::new(t_out, 0, self.quote(&input.amount, &r_in, &r_out))
+    }
+
+    fn update_safe_state_from_context(&self, ctx: &dyn Context<Self::Api>) {
+        self.update_safe_state(
+            ctx.get_first_token_reserve(),
+            ctx.get_second_token_reserve(),
+        )
+    }
+
+    fn update_safe_state_on_the_fly(&self) {
+        self.update_safe_state(
+            &self.pair_reserve(&self.first_token_id().get()).get(),
+            &self.pair_reserve(&self.second_token_id().get()).get(),
+        );
+    }
+
+    fn update_safe_state(&self, first_token_reserve: &BigUint, second_token_reserve: &BigUint) {
```

Comment:
This update has to be called first - while processing the transaction, in order to ensure that no swap was done before updateing the price.

Reply:
done